# Advent of Code

## 2017-012-007
## 2017 007

https://adventofcode.com/2017/day/7

In [2]:
# Load the input for the tower structure from the file
with open('input.txt', 'r') as file:
    tower_data = file.readlines()

# Function to determine the bottom program
def find_bottom_program(tower_data):
    # Extract all program names and the names of programs they hold
    all_programs = set()
    held_programs = set()
    
    for line in tower_data:
        parts = line.split("->")
        program_name = parts[0].split()[0]
        all_programs.add(program_name)
        if len(parts) > 1:
            held_programs.update(parts[1].strip().split(", "))
    
    # The bottom program is the one that is not held by any other program
    bottom_program = all_programs - held_programs
    return bottom_program.pop()

# Find the bottom program for the provided input
bottom_program = find_bottom_program(tower_data)
bottom_program

'dtacyn'

In [3]:
from collections import defaultdict

# Parse the input into a usable structure
def parse_tower(tower_data):
    weights = {}
    children = defaultdict(list)
    
    for line in tower_data:
        parts = line.split("->")
        program_name = parts[0].split()[0]
        program_weight = int(parts[0].split()[1].strip("()"))
        weights[program_name] = program_weight
        
        if len(parts) > 1:
            child_programs = parts[1].strip().split(", ")
            children[program_name].extend(child_programs)
    
    return weights, children

# Recursive function to calculate the total weight of a tower
def calculate_tower_weight(node, weights, children):
    return weights[node] + sum(calculate_tower_weight(child, weights, children) for child in children[node])

# Recursive function to find the unbalanced program
def find_unbalanced_program(node, weights, children):
    child_weights = {child: calculate_tower_weight(child, weights, children) for child in children[node]}
    
    if len(set(child_weights.values())) <= 1:
        # If all child towers are balanced, this node is the unbalanced one
        return None, 0
    
    # Find the unbalanced child and correct weight
    weight_counts = defaultdict(list)
    for child, weight in child_weights.items():
        weight_counts[weight].append(child)
    
    # The balanced weight is the one with more than one child
    balanced_weight = [w for w, nodes in weight_counts.items() if len(nodes) > 1][0]
    unbalanced_weight = [w for w in weight_counts if w != balanced_weight][0]
    
    unbalanced_child = weight_counts[unbalanced_weight][0]
    result, correction = find_unbalanced_program(unbalanced_child, weights, children)
    
    if result is None:
        # The unbalanced node is this child, so calculate the adjustment
        correct_weight = weights[unbalanced_child] + (balanced_weight - unbalanced_weight)
        return unbalanced_child, correct_weight
    
    return result, correction

# Parse the input and find the bottom program
weights, children = parse_tower(tower_data)
bottom_program = find_bottom_program(tower_data)

# Find the unbalanced program and its corrected weight
unbalanced_program, corrected_weight = find_unbalanced_program(bottom_program, weights, children)
unbalanced_program, corrected_weight


('ptshtrn', 521)

The program causing the imbalance is ptshtrn, and its weight needs to be adjusted to 521 to balance the entire tower.